File: Juego.py
from abc import ABC, abstractmethod
import random

# CLASE ABSTRACTA
class Personaje(ABC):
    def __init__(self, nombre):
        self.nombre = nombre
        self.vida = 100
        self.monedas = 0

    def recoger_moneda(self):
        self.monedas += 1
        print(f"{self.nombre} recogió una moneda 💰 Total: {self.monedas}")

    def recibir_dano(self, dano):
        self.vida -= dano
        if self.vida < 0:
            self.vida = 0

    def esta_vivo(self):
        return self.vida > 0

    @abstractmethod
    def atacar(self, enemigo):
        pass


# MARIO
class Mario(Personaje):
    def __init__(self):
        super().__init__("Mario")
        self.tiene_estrella = False

    def usar_poder(self):
        self.tiene_estrella = True
        print("⭐ Mario es invencible!")

    def atacar(self, enemigo):
        if self.tiene_estrella:
            print("Mario destruye al enemigo ⭐")
            enemigo.recibir_dano(100)
        else:
            print("Mario salta sobre el enemigo 🦘")
            enemigo.recibir_dano(30)


# LUIGI
class Luigi(Personaje):
    def __init__(self):
        super().__init__("Luigi")

    def atacar(self, enemigo):
        print("Luigi salta más alto 🟢")
        enemigo.recibir_dano(25)


# ENEMIGO
class Enemigo:
    def __init__(self, tipo):
        self.tipo = tipo
        self.vida = 60

    def recibir_dano(self, dano):
        self.vida -= dano
        if self.vida < 0:
            self.vida = 0

    def esta_vivo(self):
        return self.vida > 0


# JUEGO
jugadores = [Mario(), Luigi()]
enemigo = Enemigo("Goomba")

print("🍄 INICIA EL JUEGO DE MARIO 🍄")

for p in jugadores:
    if not enemigo.esta_vivo():
        break

    if random.choice([True, False]):
        p.recoger_moneda()

    if isinstance(p, Mario) and random.choice([True, False]):
        p.usar_poder()

    p.atacar(enemigo)
    print("Vida del enemigo:", enemigo.vida)
    print("-------------------")

print("🏁 Fin del juego")

File: Juego.java
import java.util.ArrayList;
import java.util.Random;

// CLASE PRINCIPAL (PRIMERO)
public class MarioGame {
    public static void main(String[] args) {

        ArrayList<Personaje> jugadores = new ArrayList<>();
        jugadores.add(new Mario());
        jugadores.add(new Luigi());

        Enemigo goomba = new Enemigo("Goomba");
        Random rand = new Random();

        System.out.println("🍄 INICIA EL JUEGO DE MARIO 🍄");

        int turno = 1;

        while (goomba.estaVivo()) {
            System.out.println("\n--- TURNO " + turno + " ---");

            for (Personaje p : jugadores) {

                if (!p.estaVivo()) continue;
                if (!goomba.estaVivo()) break;

                if (rand.nextBoolean()) {
                    p.recogerMoneda();
                }

                if (p instanceof PoderEspecial && rand.nextBoolean()) {
                    ((PoderEspecial) p).usarPoder();
                }

                p.atacar(goomba);

                System.out.println("Vida del " + goomba.getTipo() + ": " + goomba.getVida());

                if (goomba.estaVivo()) {
                    goomba.atacar(p);
                    System.out.println(p.getNombre() + " vida: " + p.getVida());
                }

                System.out.println("-------------------");
            }

            turno++;
        }

        System.out.println("🏁 Fin del juego");
    }
}

// INTERFACE
interface PoderEspecial {
    void usarPoder();
}

// CLASE ABSTRACTA
abstract class Personaje {
    protected String nombre;
    protected int vida;
    protected int monedas;

    public Personaje(String nombre) {
        this.nombre = nombre;
        this.vida = 100;
        this.monedas = 0;
    }

    public void recogerMoneda() {
        monedas++;
        System.out.println(nombre + " recogió una moneda 💰 Total: " + monedas);
    }

    public void recibirDano(int dano) {
        vida -= dano;
        if (vida < 0) vida = 0;
    }

    public boolean estaVivo() {
        return vida > 0;
    }

    public String getNombre() {
        return nombre;
    }

    public int getVida() {
        return vida;
    }

    public abstract void atacar(Enemigo e);
}

// MARIO
class Mario extends Personaje implements PoderEspecial {
    private int turnosEstrella = 0;

    public Mario() {
        super("Mario");
    }

    public void activarEstrella() {
        turnosEstrella = 2;
        System.out.println("⭐ Mario es invencible por 2 turnos!");
    }

    @Override
    public void usarPoder() {
        activarEstrella();
    }

    @Override
    public void atacar(Enemigo e) {
        if (turnosEstrella > 0) {
            System.out.println("⭐ Mario destruye al enemigo!");
            e.recibirDano(50);
            turnosEstrella--;
        } else {
            System.out.println("Mario salta sobre el enemigo 🦘");
            e.recibirDano(20);
        }
    }
}

// LUIGI
class Luigi extends Personaje {

    public Luigi() {
        super("Luigi");
    }

    @Override
    public void atacar(Enemigo e) {
        System.out.println("Luigi salta más alto 🟢");
        e.recibirDano(25);
    }
}

// ENEMIGO
class Enemigo {
    private String tipo;
    private int vida;

    public Enemigo(String tipo) {
        this.tipo = tipo;
        this.vida = 80;
    }

    public void recibirDano(int dano) {
        vida -= dano;
        if (vida < 0) vida = 0;
    }

    public void atacar(Personaje p) {
        System.out.println(tipo + " ataca a " + p.getNombre() + " 👾");
        p.recibirDano(15);
    }

    public boolean estaVivo() {
        return vida > 0;
    }

    public String getTipo() {
        return tipo;
    }

    public int getVida() {
        return vida;
    }
}
